client/state/utils.js

```diff
@@ -253,7 +272,7 @@ export const withSchemaValidation = ( schema, reducer ) => {
 	}
 
 	const wrappedReducer = ( state, action ) => {
-		if ( action.type === DESERIALIZE && ! ( state && isValidStateWithSchema( state, schema ) ) ) {
+		if ( action.type === DESERIALIZE && ! isValidSerializedState( schema, reducer, state ) ) {
```

Comment:
Why not leave the check for `state` here? It could bypass a function call. Would want to make it check for `undefined` if you do though

Reply:
That's a good idea. I moved the `state === undefined` check outside the `isValidSerializedState`, as its purpose is somewhat different than the rest of the code. And added some comments about some very subtle gotchas that are not visible at first sight (and not even at the second :smile:).

Note that keeping the `! state` check would incorrectly ignore a stored state of a reducer with single boolean value `false` or numeric value `0` or empty string or...